Key filter

Hublerho opened the pull request:
This PR should implement the feature to use key based filtering of auditd events.
Specifying keys in laurel.conf:

[filter]
filter-keys = ["filter-this"]

#24 Currently it only drops the whole message

Looking forward for your feedback

hillu commented:
Thank you for your PR!

Could you please rework your commit series so that there are at most a handful of self-contained, linear commits to review?

The `pub filter: bool` you added to the `Coalesce` object should really be part of the `Event`.

Hublerho commented:
Hi,
I reworked the series and added the `filter: bool` to the event.
Assuming an event should not be filtered upon creation the default is set to false.
Only while processing the value will be set accordingly.

I hope this resolves part of the issue.
Excited to hear from you!

hillu commented on src/config.rs:
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
         let c: Config = toml::de::from_str(r#"
 user = "somebody"
-directory = "/path/to/somewhere"
+directory = "/etc/laurel/config.toml"
```
I think there's a misconception here. directory is meant to be the the "working directory", i.e. the base directory for log files.

hillu commented on src/config.rs:
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
         println!("{:#?}", &c);
         assert_eq!(c.user, Some("somebody".to_string()));
-        assert_eq!(c.directory, Some(Path::new("/path/to/somewhere").to_path_buf()));
+        assert_eq!(c.directory, Some(Path::new("/etc/laurel/config.toml").to_path_buf()));
```
see above

hillu commented on src/config.rs:
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
+// Run if you want to see wether or not your config file is parsed correctly
+    #[test]
+    fn simple_from_file() {
+        let config_file = fs::read_to_string("/etc/laurel/config.toml");
```
This is the line that broke the CI test.

If you feel it's necessary to parse content from a file, please include it in the `testdata` subdirectory. Most likely, howver, you should be able to get away with parsing from an inline string (see `simple` test).

hillu commented on src/main.rs:
```diff
@@ -262,6 +262,10 @@ fn run_app() -> Result<(), Box<dyn Error>> {
 
     coalesce.populate_proc_table()
         .map_err(|e| format!("populate proc table: {}", e))?;
+    print!("{:?}",config.filter.filter_keys);
```
The `println!` looks like debug output that you forgot to strip.

hillu commented on src/coalesce.rs:
```diff
@@ -570,14 +580,16 @@ impl<'a> Coalesce<'a> {
         }
     }
 
+
```
Format-only changes should go into a separate commit.

hillu reviewed (changes requested):
The core of your changes look okay now. There are only a few minor issues I'd like to be changed, after that I'd be happy to accept the PR.

Please put formatting-only changes into a separate commit. They distract from the actual functionality you added.

hillu commented:
btw, would you also like to do the the `filter.filter-labels` feature (as discussed in #24)?

Hublerho commented:
I reworked/cleaned up the commits, sorry for the mess a lot of the code was due to improper cleanup before commiting (debug code).
Hope it is ok now, otherwise I am happy to correct more :D

Hublerho commented:
#24 sounds interesting, I can give it a try

hillu commented:
After rebasing onto `master` and squashing "Test implementation" and "Cleanup", I have pushed your changes.

Thank you very much for your contribution!

How would you like to be credited in the next release notes?

Hublerho commented:
Great, thank you!

You can just tag me along with the feature if this is ok?
